### ISSUE TITLE
[Proposal] Ability to add additional meta data to Queued Jobs

### ISSUE THREAD

**Issue by miscbits** [laravel/ideas#1503]
Currently when a job is dispatched, the assumption is that Laravel will handle the job. For certain use cases this might not actually be the desired behavior though. In certain cases, the Laravel app might not have necessary access to certain resources. If thats the case, then a separate microservice would have to handle the job.

This would be idea, but Laravel sends the job to a queue using PHP's [serialize](http://php.net/manual/en/function.serialize.php) function so it is not easy for other services not written in PHP to read these jobs.

My proposal is to give the ability to add additional meta data on a job that would be parsable by non-PHP apps.

My current use case for this is simply for being able to process jobs that require resources my application doesn't have access to. Passing in the job that needs to be processed to a queue that another app can also read gives me a really nice way of passing data between apps on the queue while still using the Job Dispatching. While this is my use case, 

The ability to add meta data to a job would also be useful to others. Adding tags to a job directly would allow people viewing their jobs on an admin dashboard to search through them easily. Ex. if I were viewing a list of jobs that were responsible for resizing a user's photo, and I wanted to go through and look for a particular user by something like a username, I wouldn't really be able to. This is because the user model on the job only serializes the id of that user. This means in order to do this kind of search, I'd either have to load each user model for this dashboard (n+1 problem) or do a search for a username to the database and see if their id matches any in the list (this is a single query but its still an awkward process). If instead of any of that, a user could simply do something like 

```PHP
    public function __construct(User $user)
    {
        $this->user = $user;
        $this->meta = $user->username;
    }
```

Then they would be able to search through jobs without hitting the users table at all.

I've already coded out enough in my fork of the framework though I need to add tests for the added code. https://github.com/miscbits/framework

With this, developers should be able to add a field called meta to their job and it will appear in the message sent to whatever queue is provided. From the job I created to test this
```PHP
    /**
     * Create a new job instance.
     *
     */
    public function __construct()
    {
        $this->meta = 'This is some meta data I added in my constructor';
    }
```

The code above produces below

```json
{
  "data": {
    "command": "...",
    "commandName": "App\\Jobs\\MyJob",
    "meta": "This is some meta data I added in my constructor"
  },
  "displayName": "App\\Jobs\\MyJob",
  "job": "Illuminate\\Queue\\CallQueuedHandler@call",
  "maxTries": null,
  "timeout": null,
  "timeoutAt": null
}
```

If no meta is added, then the current behavior is used. 

The two files changed to get this feature are `Illuminate\Bus\Dispatcher` and `Illuminate\Queue\Queue`. These changes broke 0 tests in the Queue tests and broke 1 test in the Dispatcher tests. That one broken test was because the test was trying to see that the method laterOne was receiving specific params, but the new code was passing in `''` for the meta since no meta was defined in the mock job. adding that as a param in the test got it back to passing.

**Comment by miscbits** [comment]
a few questions. 
1) I considered the meta a string, though with this if you pass something that can be converted to string that also works. I also considered making the meta object an array. This would just mean before being added to the data, it would have to be parsed to json. This might be a cleaner way to go about it. What do you all think about that? 

2) knowing that you would have to add a meta property to get this behavior seems unlikely and unintuitive. Is there a better way that is more in line with the current functionality to get this?

I can't really think of any. I wanted as much of this as possible to be written in the job, so I figured it wouldn't be appropriate to set this during the dispatch, ergo something like 

```PHP
MyJob::dispatch()->withMeta('someData');
```

though I did consider that. Would that be better than containing the meta in the Job itself?

**Comment by sisve** [comment]
> Currently when a job is dispatched, the assumption is that Laravel will handle the job. For certain use cases this might not actually be the desired behavior though. In certain cases, the Laravel app might not have necessary access to certain resources. If thats the case, then a separate microservice would have to handle the job.
> 
> This would be idea, but Laravel sends the job to a queue using PHP's serialize function so it is not easy for other services not written in PHP to read these jobs.
>
> [...]
>
> My current use case for this is simply for being able to process jobs that require resources my application doesn't have access to. Passing in the job that needs to be processed to a queue that another app can also read gives me a really nice way of passing data between apps on the queue while still using the Job Dispatching. [...]

You've probably only tried dispatching Laravel jobs. In that case, yes, they will be serialized and called via CallQueuedHandler@call, and the consumer get's the fun job of deserializing the data.

However, you can dispatch strings and custom data already, and end up with a payload column that is easier to use. It just isn't available via the dispatch() helper.

    \Queue::pushOn('other-microservice-queue', 'VoteForPresident', array('username' => 'sisve'));

    // payload column, json-decoded
    array(
        'displayName' => 'VoteForPresident',
        'job' => 'VoteForPresident',
        'maxTries' => null,
        'timeout' => null,
        'data' => array(
            'username' => 'sisve'
        ),
    )

> The ability to add meta data to a job would also be useful to others. Adding tags to a job directly would allow people viewing their jobs on an admin dashboard to search through them easily. Ex. if I were viewing a list of jobs that were responsible for resizing a user's photo, and I wanted to go through and look for a particular user by something like a username, I wouldn't really be able to. [...]

This really sounds like you're using the database queue. (I'm not sure how this would work in other queue systems.) Having the data inside a meta property still makes it very hard to query effectively.

But you can write a custom database queue to support this!

1. Inherit from the existing DatabaseQueue.
2. Override createPayloadArray which is responsible for creating the above array.

        protected function createPayloadArray($job, $data = '', $queue = null) {
            $array = parent::createPayloadArray($job, $data, $queue);
        
            if ($job instanceof HasUserIdInterface) {
                $array['userId'] = $job->getUserId();
            } else {
                $array['userId'] = null;
            }
        
            return $array;
        }

3. Override buildDatabaseRecord which builds the actual array used for the insert statement. This method only has access to a json-encoded version of the above payload.

        protected function buildDatabaseRecord($queue, $payload, $availableAt, $attempts = 0) {
            $array = parent::buildDatabaseRecord($queue, $payload, $availableAt, $attempts);
            $array['user_id'] = Arr::get(json_decode($payload, true), 'userId');
            return $array;
        }

4. Create the necessary migrations so that you have a user_id column in your jobs table.

This should be enough to get a jobs.user_id column populated with user information for every job. You can now easily query who every job belongs to.

    SELECT users.id, users.name,
           COUNT(jobs.*) AS num_jobs
      FROM users
     INNER JOIN jobs ON users.id = jobs.user_id
     GROUP BY users.id, users.name
    HAVING COUNT(jobs.*) > 0

(If you're bored, you can then later use this column and override getNextAvailableJob to make sure you do not pick jobs from a user that already has another job executing. That way a user can only execute one job at a time. https://github.com/laravel/ideas/issues/1405#issuecomment-442454330)

**Comment by miscbits** [comment]
Using Queue push works for this situation, but the main goal of this suggestion was to just use the dispatch syntax and keep the ability to use things like delays for dispatching.

Passed that hitting the users table in that second use case was something I was trying to avoid, especially since I am not using the database queue and it wouldn't be guaranteed that you are. In most cases you would be using a Redis queue so I was really writing examples using that. 

Basically, yes, you are correct that there are effective ways to push to a queue with custom data, but the goal of this was to do so with the dispatch helper. It was also to contain that custom data in a job rather than in a controller or service.

**Comment by themsaid** [comment]
What's the point of sending a job via dispatch() if the other service won't be able to serialize the job ? Using `Queue:pushRaw()` makes more sense in your case.

**Comment by miscbits** [comment]
@themsaid the point was that the other services don't need to serialize it because a third party source would handle the job. 

The point of this though was that I wanted to use the Event Dispatch model despite not using the Job Handlers. In the end though, the suggestion of pushRaw is in fact where we landed. Closing this as it seems that there wasn't really interest for this outside myself.

**Comment by pratikjoshibyjus** [comment]
generic post added here https://github.com/laravel/ideas/issues/2309
